07-errors.md

```diff
@@ -0,0 +1,301 @@
+# Error Conditions
+
+This document specifies the error conditions in the FVM and the associated error codes/numbers.
+
+## Exit Codes
+
+Exit codes are numeric, 32bit integer codes that are returned when an actor exits both successfully and in error.
+Exit codes appear on-chain inside the message receipt, and therefore form part of consensus.
+
+We classify exit codes as "normal" and "abnormal".
+
+The only _normal_ exit code is `0`, or `Ok`. The `Ok` exit code indicates that the message was
+applied successfully.
+
+All other exit codes are _abnormal_ and indicate that the message _was not_ applied and that the
+chain state _was not_ updated.
+
+### System "Abnormal Exit" Codes
+
+System abnormal exit codes are generated by the VM itself to indicate that the message itself failed
+to apply. These exit codes will never be observed or returned directly by actors.
```

Comment:
"Never observed" where? They are observed on chain receipts (at least SysErrOutOfGas is). If some are, and others aren't, then we need to subclassify.

Reply:
Ah, that's an ambiguity:

> These exit codes will never be (observed or returned) directly by actors.

How about:

> These exit codes will never be observed by or returned by actors.

?

Reply:
Hm, what would happen if an actor performs a send syscall to an inexistent actor? I would imagine that it would observe an `SysErrInvalidReceiver`. That should be the case to enable counterfactual cases?

Reply:
Currently, it sees an `ErrNotFound` syscall error.